### PR TITLE
Actions

### LINKED ISSUE #1000: "action" interface for mouse & keyboard input, and take screenshots

### PR DESCRIPTION
**Pull request by emaicus**
Closes #1000

Brief: Adds support for scripted keyboard and mouse interaction, screenshots, and delays for use in interactive programs. While developing, some modifications were made to execute.cpp, maincompile.cpp, etc. but most major changes take place in the added window_utils.cpp file. I have added documentation on Submitty.org under the Assignment Configuration heading (the section is called Types of Action) which specifies the interactions I have implemented and the commands needed to utilize them. 

Breakdown:
When the execute function in execute.cpp is called, the first thing that happens is a check if the instructor has specified any actions. If they have, we assume that a window will be created to use them on. The process forks as it did before, with the child running the student's code. The parent, however, begins looking for the expected window every 1/10th second (only if a window is expected, 1/10th second was the value specified before, and all memory/time checks on the student's program are still in place). If the window is found (for now, if a window is spawned with a pid matching the child's), the program begins taking actions, one every 1/10th second. While there are still actions to take, the execute function calls the takeAction function in window_utils.cpp, which routes the provided command to the proper function (e.g., if the next action is a delay of the form "delay 4", the takeAction function will parse it and forward it to the delay function). If a delay is invoked, it is broken into 1/10th second portions, with memory and time checks performed when appropriate (per the previous implementation of execute). Before any action is taken, a check is performed to see if the student's window still exists. if it does, the mouse is centered on the student's window and the window is brought into focus before the action is taken. If the student's window has failed, the action is not taken. All mouse commands should be clamped to within the student's window minus a one pixel boundary, meaning that unless the mouse is physically manipulated, it should not be able to leave the student's window and interact with the pc. 

Known Issues: 
1) Programs involving actions should be run sequentially, as more than one program can not manipulate the mouse or keyboard concurrently. However, merging this code should not have any effect on the ability of standard programs to execute in parallel. 
2) In order to allow untrusted users to access the screen, the command: xhost +SI:localuser:USER must be run, where USER is replaced by the desired user.

### CODE REVIEW

File: grading/execute.cpp [hunk 1]
```diff
@@ -103,7 +104,12 @@ bool system_program(const std::string &program, std::string &full_path_executabl
     // for Network Programming
     { "timeout",                 "/usr/bin/timeout" },
     { "mpicc.openmpi",           "/usr/bin/mpicc.openmpi" },
-    { "mpirun.openmpi",          "/usr/bin/mpirun.openmpi" }
+    { "mpirun.openmpi",          "/usr/bin/mpirun.openmpi" },
+
+    // for graphics/window interaction
+    { "scrot",                   "/usr/bin/scrot"},
+    { "xdotool",                   "/usr/bin/xdotool"}
```
Comment (bmcutler) [inline grading/execute.cpp]: indentation

File: grading/execute.cpp [hunk 2]
```diff
@@ -803,11 +810,25 @@ void TerminateProcess(float &elapsed, int childPID) {
...
+      const std::vector<std::string> actions,
+      const std::string &execute_logfile,
 	    const nlohmann::json &test_case_limits,
 	    const nlohmann::json &assignment_limits,
             const nlohmann::json &whole_config) {
 
+  bool window_mode = false; //Tells us if the process is expected to spawn a window. (additional support later) 
+  
+  if(actions.size() > 0){ //right now, we assume if there are actions, there will be a window.  
+    std::cout << "Recieved " << actions.size() << " actions" << std::endl; //useful debug line.
```
Comment (bmcutler) [inline grading/execute.cpp]: spelling

File: grading/window_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+* using a regex, which simply strips away anything before one or more tabs or spaces.
+* Runs in O(number_of_windows)
+*/
+std::vector<std::string> getWindowNameAssociatedWithPid(int pid)
+{
+  std::cout << "Attempting to find a window associated with pid: " << pid << std::endl;
+
+  /*
+  * At the moment, this method finds any window names associated with the child's pid. This works fine.
+  * However, it is concieveable that the child could fork/generate a window with its own pid. In this case
```
Comment (bmcutler) [inline grading/window_utils.cpp]: spelling

File: grading/window_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+      }
+      else{
+        continue;
+      }
+    }
+    else{
+      continue;
+    }
+
+    if(std::regex_match(window, match, regex)){ //get the third collumn (the pid, view example output above)
```
Comment (bmcutler) [inline grading/window_utils.cpp]: spelling

File: grading/window_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+      intersection_point = getLineIntersectionPoint(current_point, projected_point, lower_left, upper_left);
+      if(intersection_point.size() != 0){
+        std::cout << "intersected left" << std::endl;
+        moved_mouse_x = (int)intersection_point[0]; 
+        moved_mouse_y = (int)intersection_point[1];
+        found = true;
+      }
+    }
+
+    if(!found) //we are inside of the box (gauranteed by clamp) so move
```
Comment (bmcutler) [inline grading/window_utils.cpp]: spelling

**Comment by bmcutler** [comment]
Also, please pull master & fix merge conflicts & re-test w/ graphics & non graphics programs.

File: grading/execute.cpp [hunk 6]
```diff
@@ -106,8 +106,10 @@ bool system_program(const std::string &program, std::string &full_path_executabl
     { "mpirun.openmpi",          "/usr/bin/mpirun.openmpi" },
 
     // for graphics/window interaction
-    { "scrot",                   "/usr/bin/scrot"},
-    { "xdotool",                 "/usr/bin/xdotool"}
+    { "scrot",                   "/usr/bin/scrot"}, //screenshot utility
+    { "xdotool",                 "/usr/bin/xdotool"}, //keyboard/mouse input
+    { "xdotool",                 "/usr/bin/wmctrl"}, //bring window into focus
```
Comment (bmcutler) [inline grading/execute.cpp]: typo?  repeated program name

Comment (emaicus) [inline grading/execute.cpp]: Good catch. I just pushed an update.

**Pull request merged**